File: test/lib/cmdLineRender/cpp/testActiveRenderDescriptionSceneGlobals.cpp
// Copyright 2026 Autodesk
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "testUtils.h"

#include <pxr/imaging/hd/sceneGlobalsSchema.h>

#include <ufe/pathString.h>

#include <flowViewport/selection/fvpPathMapperRegistry.h>
#include <gtest/gtest.h>

#include <string>


PXR_NAMESPACE_USING_DIRECTIVE

namespace {

struct _ActiveRenderDescriptionPaths
{
    SdfPath renderSettings;
    SdfPath renderPass;
};

_ActiveRenderDescriptionPaths _GetActiveRenderDescriptionPaths()
{
    const auto& sceneIndices = GetTerminalSceneIndices();
    EXPECT_EQ(sceneIndices.size(), 1u)
        << "Expected one terminal scene index, but found " << sceneIndices.size();
    if (sceneIndices.size() != 1) {
        return {};
    }

    const HdSceneGlobalsSchema globals
        = HdSceneGlobalsSchema::GetFromSceneIndex(sceneIndices.front());
    EXPECT_TRUE(globals.IsDefined());
    if (!globals.IsDefined()) {
        return {};
    }

    const HdPathDataSourceHandle renderSettings = globals.GetActiveRenderSettingsPrim();
    const HdPathDataSourceHandle renderPass = globals.GetActiveRenderPassPrim();
    EXPECT_TRUE(renderSettings);
    EXPECT_TRUE(renderPass);
    if (!renderSettings || !renderPass) {
        return {};
    }

    return {
        renderSettings->GetTypedValue(0.0f),
        renderPass->GetTypedValue(0.0f),
    };
}

SdfPath _GetSceneIndexPath(const char* usdPrimPath)
{
    const Ufe::Path appPath
        = Ufe::PathString::path(std::string("|renderSettings|renderSettingsShape,") + usdPrimPath);
    const SdfPath result = Fvp::sceneIndexPath(appPath);
    EXPECT_FALSE(result.IsEmpty());
    return result;
}

SdfPath _GetDefaultRenderSettingsPath()
{
    const Ufe::Path appPath
        = Ufe::PathString::path("UsdDefaultRenderDescription,/Render/SceneRenderSettings");
    const SdfPath result = Fvp::sceneIndexPath(appPath);
    EXPECT_FALSE(result.IsEmpty());
    return result;
}

} // namespace

TEST(TestActiveRenderDescriptionSceneGlobals, RenderSettings)
{
    const auto active = _GetActiveRenderDescriptionPaths();

    EXPECT_EQ(active.renderSettings, _GetSceneIndexPath("/Render/MainRender"));
    EXPECT_TRUE(active.renderPass.IsEmpty());
}

TEST(TestActiveRenderDescriptionSceneGlobals, ValidRenderPass)
{
    const auto active = _GetActiveRenderDescriptionPaths();

    EXPECT_EQ(active.renderSettings, _GetSceneIndexPath("/Render/MainRender"));
    EXPECT_EQ(active.renderPass, _GetSceneIndexPath("/Render/Passes/foreground"));
}

TEST(TestActiveRenderDescriptionSceneGlobals, InvalidRenderPassSource)
{
    const auto active = _GetActiveRenderDescriptionPaths();

    EXPECT_EQ(active.renderSettings, _GetDefaultRenderSettingsPath());
    EXPECT_EQ(active.renderPass, _GetSceneIndexPath("/Render/Passes/foreground"));
}

TEST(TestActiveRenderDescriptionSceneGlobals, MissingRenderPass)
{
    const auto active = _GetActiveRenderDescriptionPaths();

    EXPECT_EQ(active.renderSettings, _GetDefaultRenderSettingsPath());
    EXPECT_TRUE(active.renderPass.IsEmpty());
}

File: test/lib/cmdLineRender/testActiveRenderDescriptionSceneGlobals.py
# Copyright 2026 Autodesk, Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import shutil

import maya.cmds as cmds
import fixturesUtils
import mtohUtils

from testUtils import PluginLoaded


_SCENES_DIR = os.path.join(
    os.path.dirname(os.path.abspath(__file__)),
    "scenes", "renderSettings")


def _scenePath(sceneFileName):
    return os.path.join(_SCENES_DIR, sceneFileName)


class TestActiveRenderDescriptionSceneGlobals(mtohUtils.MayaHydraBaseTestCase):
    _file = __file__
    _initializeStandalone = True
    _setHdStormRenderer = False
    _requiredPlugins = ["mtoa"]

    def _runBranch(self, sceneFileName, testCase):
        tempAppDir = os.environ.get("MAYA_APP_DIR")
        if tempAppDir:
            destDir = os.path.join(tempAppDir, "projects", "default", "scenes")
            if not os.path.isdir(destDir):
                os.makedirs(destDir)
            baseName = os.path.splitext(sceneFileName)[0]
            for ext in (".ma", ".usda"):
                shutil.copy(_scenePath(baseName + ext), os.path.join(destDir, baseName + ext))
            scenePath = os.path.join(destDir, sceneFileName)
        else:
            scenePath = _scenePath(sceneFileName)

        cmds.file(scenePath, open=True, force=True)
        cmds.hydraRender(renderer="HdArnoldRendererPlugin")

        try:
            with PluginLoaded("mayaHydraCppTests"):
                cmds.mayaHydraCppTest(
                    f="TestActiveRenderDescriptionSceneGlobals.%s" % testCase)
        finally:
            cmds.mayaHydraTesting(releaseBatchRenderer=True)

    def test_RenderSettings(self):
        self._runBranch("activeRenderDescriptionRenderSettings.ma", "RenderSettings")

    def test_ValidRenderPass(self):
        self._runBranch("activeRenderDescriptionValidPass.ma", "ValidRenderPass")

    def test_InvalidRenderPassSource(self):
        self._runBranch(
            "activeRenderDescriptionInvalidPassSource.ma", "InvalidRenderPassSource")

    def test_MissingRenderPass(self):
        self._runBranch("activeRenderDescriptionMissingPass.ma", "MissingRenderPass")


if __name__ == "__main__":
    fixturesUtils.runTests(globals())
